imap_processing/ialirt/process_ephemeris.py

```diff
@@ -144,20 +121,77 @@ def calculate_azimuth_and_elevation(
             azccw=False,  # Azimuth measured clockwise from the positive y-axis
             elplsz=True,  # Elevation increases from the XY plane toward +Z
-            obspos=observer_position_ecef,  # observer pos. to center of motion
+            obspos=ground_station_position_ecef,  # observer pos. to center of motion
             obsctr="EARTH",  # Name of the center of motion
             obsref="IAU_EARTH",  # Body-fixed, body-centered reference frame wrt
             # observer's center
         )
         azimuth.append(np.rad2deg(azel_results[0][1]))
         elevation.append(np.rad2deg(azel_results[0][2]))
 
-    # TODO: potentially use the velocity components returned from azlcpo to
-    # TODO: calculate doppler
-
     return np.asarray(azimuth), np.asarray(elevation)
 
 
+@typing.no_type_check
+def calculate_doppler(
+    longitude: float,
+    latitude: float,
+    altitude: float,
+    observation_time: Union[float, np.ndarray],
+) -> Union[float, ndarray[float]]:
+    """
+    Calculate the doppler velocity.
+
+    Notes about the spkezr function (wrapped in imap_state):
+    The function returns the state of the target (state) and the light time.
+    The first three components of state represent the x-, y- and z-components of the
+    target's position; the last three components form the corresponding velocity vector.
+
+    Parameters
+    ----------
+    longitude : float
+        Longitude in decimal degrees. Positive east of prime meridian,
+        negative to west.
+    latitude : float
+        Latitude in decimal degrees. Positive north of equator, negative
+        to south.
+    altitude : float
+        Altitude in kilometers.
+    observation_time : float or np.ndarray
+        Time at which the state of the target relative to the observer
+        is to be computed. Expressed as ephemeris time, seconds past J2000 TDB.
+
+    Returns
+    -------
+    doppler : float or np.ndarray[float]
+        Doppler velocity in kilometers per second.
+    """
+    ground_station_position_ecef = latitude_longitude_to_ecef(
+        longitude, latitude, altitude
+    )
+
+    # find position and velocity relative to the center of the earth using spice spkezr
+    # https://spiceypy.readthedocs.io/en/main/documentation.html#spiceypy.spiceypy.spkezr
+    state = imap_state(
+        et=observation_time, ref_frame="ITRF93", abcorr="LT+S", observer=SpiceBody.EARTH
+    )
+    # shifting position by subtracting ground station location relative to the center
+    # of the earth
+    state = state - np.pad(ground_station_position_ecef, (0, 3), "constant")
+    ax = int(state.ndim == 2)
+    # calculate radial velocity
+    if isinstance(observation_time, np.ndarray):
+        doppler = np.sum(state[:, 3:6] * state[:, 0:3], axis=ax) / np.linalg.norm(
+            state[:, 0:3], axis=ax
+        )
+    else:
+        doppler = np.sum(state[3:6] * state[0:3], axis=ax) / np.linalg.norm(
+            state[0:3], axis=ax
+        )
```

Comment:
I think that the ellipsis slicing also works here.
```suggestion
    doppler = np.sum(state[..., 3:6] * state[..., 0:3], axis=-1) / np.linalg.norm(
            state[..., 0:3], axis=-1
        )

```